keps/sig-storage/3314-csi-volume-snapshot-delta/README.md

```diff
@@ -0,0 +1,1480 @@
+[supported limits].
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+-->
+
+The Kubernetes API server proxies the CBT payloads between the user and the
+aggregated API server. See the [Risks and Mitigations](#risks-and-mitigations)
+section on using flow control to protect the Kubernetes API server.
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider, and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+The previous alternate design which involves generating and returning a callback
+endpoint to the caller has been superceded by the aggregation extension
+mechanism described in this KEP. The aggregation extension design provides a
+tighter integration with the Kubernetes API server, enabling the re-use of the
+existing Kubernetes machinery of GVR and GVK binding, URL registration and
+delegated authentication and authorization.
```

Comment:
I talked to @deads2k about possibility of a backup software talking directly to CSI sidecar using a service (when the backup SW is inside a cluster) or using a public route / ingress when it's outside of the cluster. I.e. we won't need the aggregated API server at all. The SW can read `VolumeSnapshotDeltaServiceConfiguration` CR by itself to find the sidecar. The actual protocol between the sidecar and the backup SW is IMO implementation detail, but it *can* re-use Kubernetes API authentication using e.g. kube-rbac-proxy.

This is close tho this proposal, just with the same authn/authz as the API server and CSI sidecar announcing its http(s) endpoint in VolumeSnapshotDeltaServiceConfiguration.

It all depends on size of a regular and maximum list of changed blocks - if it's O(MiB), I'd be fine with this KEP, if it's O(GiB) then while the aggregated API server looks more Kubernetes-ish, it will just eat CPU and network bandwidth and it makes the overall architecture more fragile.

/lgtm cancel

I let @msau42 judge if it's good for alpha as it is.

Reply:
Note that `/pod/logs` (and `/exec` and `/portforward`) is a kube subresource that can be `O(GiB)` but the traffic that flows across it is explicitly intended to be management traffic, and even then there have been multiple efforts to decouple exec/logs/portforward.  I agree with @deads2k and @jsafrane - it is not appropriate to design core APIs that funnel large amounts (~`O(GiB)`) of non-management traffic through aggregated APIs (third parties may do this if they so choose).  If a hypothetical decoupling layer were to be designed and implemented or the amount of data is `O(MiB)` it is less reasonable

Reply:
@jsafrane @smarterclayton thanks for the feedback. The idea of the aggregated API was definitely inspired by `/pod/logs` and `metrics-server`, where we assumed that our traffic won't be worse than theirs. Will adding AFP flow control and priority level configuration to protect the K8s api server helps here, as it proxies traffic to-and-forth our out-of-tree agg api server? Or is AFP also meant for management traffic? The implication is that cluster admin will be able to control the changed block traffic, which may or may not be desirable.

If aggregated API server isn't an appropriate choice, will SIG architecture, SIG app or both, be the right forums for me to bring this design problem to? We have explore multiple alternatives over the past few months, including:

* a CRD approach where the CSI driver returns an "out-of-band" data endpoint (via `status` subresource) to the user
* a CRD approach where we let user specify a "callback URL" for the CSI driver to deliver the data to
* direct HTTP calls to the CSI driver (without kube-style API)
* aggregated API server